workload/tpch: implement TPCH initial data

danhhz opened the pull request:
This is not spec compliant but it's intended to be close enough to the
spirit of the spec that it can be used for convenience. The spec is very
specific that official results need to use the exact output of the dbgen
program, so any real testing should use our pregenerated dbgen csv
files.

Note that dbgen is totally deterministic and the rng is already set up
to be started at non-zero row offsets, so there's no reason we can't
make this exactly identical to dbegn. It will just be fiddly.

Still TODO in followup PRs:
- Figure out how to create a small subset of the data so
  TestAllRegisteredImportFixture/tpch can run.
- Figure out how to get the date fields to work with InsertsDataLoader
  without being too intrusive to workload.

Release note: None

jordanlewis committed:
workload/tpch: implement TPCH initial data

This is not spec compliant but it's intended to be close enough to the
spirit of the spec that it can be used for convenience. The spec is very
specific that official results need to use the exact output of the dbgen
program, so any real testing should use our pregenerated dbgen csv
files.

Note that dbgen is totally deterministic and the rng is already set up
to be started at non-zero row offsets, so there's no reason we can't
make this exactly identical to dbegn. It will just be fiddly.

Still TODO in followup PRs:
- Figure out how to create a small subset of the data so
  TestAllRegisteredImportFixture/tpch can run.
- Figure out how to get the date fields to work with InsertsDataLoader
  without being too intrusive to workload.

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/42124)
<!-- Reviewable:end -->

jordanlewis reviewed (commented):
I forget what's new in this vs. what I worked on, but seems like you added a cheap text pool, so that's great!

What's left to do? (I see the PR is marked as draft)

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/42124#-:-Lt2xRtv771EfL7rR-aH:bq17vvc)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @jordanlewis)



<!-- Sent from Reviewable.io -->

danhhz commented:
Things I've changed since your branch:
- resolve rebase conflicts
- cheap text pool
- add the `--initial-data-subset` to make it easier to test this thing (e.g. the deterministic initial data output. i think this will have to be tweaked though, since the FKs won't validate when we add them
- I removed the date related hacks we added to make this work under `workload init`, need to add an error somewhere about this in case someone tries it
- I think that's it

Draft because this is where I got to by the time I had to put it down last Friday. Remaining work is above TODOs, FKs, plus a general cleanup pass. I'd like to do a performance pass in a followup PR, but dunno when/if I'll get to that.

Mind taking a look this week to see if there's anything else high-level you think we'd need before this became something we'd recommend as an option to people?

danhhz commented:
Okay, I think this is ready to go!

Since last week:
- I went ahead and did the perf pass, you'd already done most of it
- fks
- fixed some some small data generation issues preventing fks from working. ex:`%(s+1)` vs `%s+1` when generating ps_suppkey
- ripped out the short stuff so skipped this in the FixturesImport test since it takes too long. i think i see how to do it but it's a little more invasive than i was hoping
- fixed copyright headers (lol this was an old branch)
- fixed lint errors

jordanlewis reviewed (commented):
I think the main issue with this remaining is that there are tests that check the actual TPCH sample queries, and their results. My understanding is that this method of generation doesn't produce exactly the same data as the "official" dbgen program, right? That would definitely preclude usage in our tests.

But if there are non test use cases that are good with this then I think we should merge it!

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/42124#-:-M-keIJi9vpkntOzocX6:boemk7k)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @danhhz)

___
*[pkg/workload/tpch/generate.go, line 1 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/42124#-M-ke89nDKy_VOwkR3Ox:-M-ke89nDKy_VOwkR3Oy:b2ni78k) ([raw file](https://github.com/cockroachdb/cockroach/blob/468e6a80780877c65d7b7718d32f1235a4283275/pkg/workload/tpch/generate.go#L1)):*
> ```Go
> // Copyright 2019 The Cockroach Authors.
> ```

lol


<!-- Sent from Reviewable.io -->

danhhz commented:
Thanks for taking a look!

> I think the main issue with this remaining is that there are tests that check the actual TPCH sample queries, and their results. My understanding is that this method of generation doesn't produce exactly the same data as the "official" dbgen program, right? That would definitely preclude usage in our tests.
> But if there are non test use cases that are good with this then I think we should merge it!

Yeah, this is totally doable (I looked at the dbgen source and the rng is straightforward to port) but a big lift. I think we should do it in two stages to prevent this from rotting further. Merge this, then do that as a followup. Are you okay with that? If so, I'll rebase this one last time and get it in.

jordanlewis commented:
Yep that sounds good to me!

On Thu, Feb 13, 2020 at 1:05 PM Daniel Harrison <notifications@github.com>
wrote:

> Thanks for taking a look!
>
> I think the main issue with this remaining is that there are tests that
> check the actual TPCH sample queries, and their results. My understanding
> is that this method of generation doesn't produce exactly the same data as
> the "official" dbgen program, right? That would definitely preclude usage
> in our tests.
> But if there are non test use cases that are good with this then I think
> we should merge it!
>
> Yeah, this is totally doable (I looked at the dbgen source and the rng is
> straightforward to port) but a big lift. I think we should do it in two
> stages to prevent this from rotting further. Merge this, then do that as a
> followup. Are you okay with that? If so, I'll rebase this one last time and
> get it in.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/cockroachdb/cockroach/pull/42124?email_source=notifications&email_token=AAAKWLP3O4NJ2RGZYZ4V7UDRCWDUXA5CNFSM4JH7RBE2YY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOELWAKRY#issuecomment-585893191>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAAKWLLNI27NQQTYMUASFRDRCWDUXANCNFSM4JH7RBEQ>
> .
>

danhhz commented:
TFTR!

bors r=jordanlewis
